hooks/post-command

```diff
@@ -19,23 +18,42 @@ do
       filename=${filename::${#filename}-1}
     fi
 
-    echo "--- compressing ${path}"
-    if [[ $OSTYPE != "msys" ]] ; then
-      sudo tar -czf ${filename}.tar.gz ${path}	
-    else	
-      tar -czf ${filename}.tar.gz ${path}	
+    if [[ $OSTYPE != "msys" ]]; then
+      sudo tar -cf - ${path} | gzip --no-name > ${filename}.tar.gz
+    else
+      tar -cf - ${path} | gzip --no-name > ${filename}.tar.gz
     fi
 
+    md5localobject=$(openssl md5 ${filename}.tar.gz |base64)
+
     # remove '/' as it breaks S3 pathing
     # remove ' ' as it breaks aws s3 cp command
     label=${BUILDKITE_LABEL//\//}
     label=${label// /_}
 
     s3_bucket="s3://${BUILDKITE_PLUGIN_CACHE_S3_BUCKET}/${BUILDKITE_PIPELINE_SLUG}/${label}"
 
-    echo "--- cp ${filename}.tar.gz into :aws: :amazon-s3:"
-    aws s3 cp "${filename}.tar.gz" "${s3_bucket}/${filename}.tar.gz"
+    # check if the tar file exists
+    if aws s3 ls ${s3_bucket}/${filename}.tar.gz; then
+      echo "comparing checksums..."
+      md5s3object=$(aws s3api head-object --bucket ${BUILDKITE_PLUGIN_CACHE_S3_BUCKET} --key ${BUILDKITE_PIPELINE_SLUG}/${label}/${filename}.tar.gz | jq -r '.Metadata.md5checksum')
+
+      if [[ ${md5s3object} != "null" ]]; then
+        if [[ ${md5s3object} == ${md5localobject} ]]; then
+          echo "skipping upload, checksums are identical"
+        else
+          echo "copying cache into s3"
+          aws s3 cp ${filename}.tar.gz ${s3_bucket}/${filename}.tar.gz --metadata md5checksum=${md5localobject}
+        fi
+      else
+        echo "copying cache into s3"
+        aws s3 cp ${filename}.tar.gz ${s3_bucket}/${filename}.tar.gz --metadata md5checksum=${md5localobject}
+      fi
```

Comment:
I think this logic could be simplified a bit to reduce the repetition of "copying cache into s3"

```suggestion
      if [[ ${md5s3object} != "null" ]] && [[ ${md5s3object} == ${md5localobject} ]]; then
        echo "skipping upload, checksums are identical"
      else
        echo "copying cache into s3"
        aws s3 cp ${filename}.tar.gz ${s3_bucket}/${filename}.tar.gz --metadata md5checksum=${md5localobject}
      fi
```